### DISCUSSION TITLE
Why must default-terminal be ‘screen’, ‘tmux’ or a derivative of them? Is this still true today?

### DISCUSSION THREAD

**Discussion by its-kostya** [tmux/tmux#4283] [General]
`tmux(1)` states 
```
default-terminal terminal
       Set the default terminal for new windows created in this session -
       the default value of the TERM environment variable.  For tmux to
       work correctly, this must be set to ‘screen’, ‘tmux’ or a deriva‐
       tive of them.
```
It is my understanding that `$TERM` informs applications about the capabilities of the terminal which can affect various behaviors such as color support, control sequences, function keys and mouse support, and other terminal features.

`terminfo` is a database and a set of libraries to describe the capabilities and characteristics of various terminal types. Applications lookup `$TERM` in terminfo db and can dynamically adjust their output based on the terminal's capabilities.

I'd like to better understand, say there is a program call `foo`. When `foo` starts, it will use the env variable `$TERM` to determine the terminal and then look up the terminal's terminfo. Here is an `strace` from `foo`:

```
newfstatat(AT_FDCWD, "/home/user/.terminfo", 0x58d4fbc6a020, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/etc/terminfo", {st_mode=S_IFDIR|0755, st_size=4096, ...}, 0) = 0
newfstatat(AT_FDCWD, "/lib/terminfo", {st_mode=S_IFDIR|0755, st_size=4096, ...}, 0) = 0
newfstatat(AT_FDCWD, "/usr/share/terminfo", {st_mode=S_IFDIR|0755, st_size=4096, ...}, 0) = 0
access("/etc/terminfo/a/alacritty", R_OK) = 0
openat(AT_FDCWD, "/etc/terminfo/a/alacritty", O_RDONLY) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=3652, ...}, AT_EMPTY_PATH) = 0
read(3, "\32\1&\0&\0\17\0\235\1\n\6alacritty|alacritty "..., 32768) = 3652
read(3, "", 28672)                      = 0
close(3)
```

If I am using a terminal like `alacritty`, `foo` wants `alacritty`'s terminfo. But if I launch `tmux` from my `alacritty` terminal, doesn't `tmux` make more instances of `alacritty`? Shouldn't `foo` want `alacritty`'s terminfo, or `tmux`'s? 

Or, does `foo` read `tmux-256color` terminfo so it knows how to output into `tmux` and `tmux` reads `alacritty`'s terminfo so that it can output to `alacritty`? So an output from `foo` will be `foo > tmux > alacritty`.

The capabilities between `tmux-256color` and `alacritty` and pretty significant and I am wondering if I cannot set `default-terminal` to be `alacritty` if I will be giving up some capabilities. 

```
user ~$ infocmp -d tmux-256color alacritty
comparing tmux-256color to alacritty.
    comparing booleans.
        bce: F:T.
        ccc: F:T.
        mc5i: F:T.
        npc: F:T.
    comparing numbers.
        pairs: 65536, 32767.
    comparing strings.
        acsc: '++\,\,--..00``aaffgghhiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz{{||}}~~', '``aaffggiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz{{||}}~~'.
        blink: '\E[5m', NULL.
        clear: '\E[H\E[J', '\E[H\E[2J'.
        cnorm: '\E[34h\E[?25h', '\E[?12l\E[?25h'.
        cuu1: '\EM', '\E[A'.
        cvvis: '\E[34l', '\E[?12;25h'.
        dsl: '\E]0;\007', '\E]2;\007'.
        ech: NULL, '\E[%p1%dX'.
        enacs: '\E(B\E)0', NULL.
        flash: '\Eg', '\E[?5h$<100/>\E[?5l'.
        initc: NULL, '\E]4;%p1%d;rgb:%p2%{255}%*%{1000}%/%2.2X/%p3%{255}%*%{1000}%/%2.2X/%p4%{255}%*%{1000}%/%2.2X\E\\'.
        is2: '\E)0', '\E[!p\E[?3;4l\E[4l\E>'.
        kb2: NULL, '\EOE'.
        kend: '\E[4~', '\EOF'.
        kent: NULL, '\EOM'.
        khome: '\E[1~', '\EOH'.
        mc0: NULL, '\E[i'.
        mc4: NULL, '\E[4i'.
        mc5: NULL, '\E[5i'.
        nel: '\EE', NULL.
        oc: NULL, '\E]104\007'.
        rep: NULL, '%p1%c\E[%p2%{1}%-%db'.
        rin: NULL, '\E[%p1%dT'.
        rmacs: '^O', '\E(B'.
        rmam: NULL, '\E[?7l'.
        rmcup: '\E[?1049l', '\E[?1049l\E[23;0;0t'.
        rmm: NULL, '\E[?1034l'.
        rs1: NULL, '\Ec\E]104\007'.
        rs2: '\Ec\E[?1000l\E[?25h', '\E[!p\E[?3;4l\E[4l\E>'.
        sgr: '\E[0%?%p6%t;1%;%?%p2%t;4%;%?%p1%p3%|%t;7%;%?%p4%t;5%;%?%p5%t;2%;%?%p7%t;8%;m%?%p9%t\016%e\017%;', '%?%p9%t\E(0%e\E(B%;\E[0%?%p6%t;1%;%?%p5%t;2%;%?%p2%t;4%;%?%p1%p3%|%t;7%;%?%p4%t;5%;%?%p7%t;8%;m'.
        sgr0: '\E[m\017', '\E(B\E[m'.
        smacs: '^N', '\E(0'.
        smam: NULL, '\E[?7h'.
        smcup: '\E[?1049h', '\E[?1049h\E[22;0;0t'.
        smm: NULL, '\E[?1034h'.
        tsl: '\E]0;', '\E]2;'.
        u8: '\E[?1;2c', '\E[?%[;0123456789]c'.
```
Perhaps I am over complicating things, but I am trying to understand how things work. I'd like to know why I need the following in my `tmux.conf`
```
set -g default-terminal "tmux-256color"
set -ag terminal-overrides ",$TERM:RGB"
```
instead of

```
set -g default-terminal "$TERM"
set -ag terminal-overrides ",$TERM:RGB"
```

**Comment by nicm** [comment]
I'm not sure how you can write your first two paragraphs and not know why TERM needs to be set correctly. If TERM tells the application to clear the terminal by sending `foo` but the terminal expects to receive `bar` to clear, obviously the application will not be able to clear the terminal and may in fact be asking the terminal to do something entirely different when it tries.

Setting TERM to alacritty inside tmux is likely to mostly work because tmux and alacritty have many commonalties and are both based on the protocol originally designed for the DEC VT100 and later DEC terminals. But you may find that some features of some applications do not work, or work incorrectly.

**Reply by its-kostya** [reply]
Well, I'm still learning and I have a knowledge gap which I am trying to close.

What trips me up is the introduction of another layer of abstraction between what is displayed and the running application (say a text editor). If the application wants to flash the screen and tmux expects `foo` but alacritty expects `bar`, what does the application send? I expect the application sends `foo` which is interpreted by tmux and then tmux send `bar` to alacritty?

**Reply by nicm** [reply]
Right. tmux is a terminal that happens to run inside another terminal.

**Reply by its-kostya** [reply]
Awesome, I think I have enough of a mental model of how this works - thanks!

I've also dug up that default-terminal `screen` _and_ `tmux` are acceptable because `tmux` is derived from `screen` : https://invisible-island.net/ncurses/terminfo.src.html#t20150502 ; cool! 

Kinda of crazy that tmux's terminal description has been part of the terminal database since _only_ 2015. I thought it was way more OG; certainly feels that way due to its ubiquity.
